### PR TITLE
GH-460: As a smith I want to add support for definition site variance in constraint solver

### PR DESCRIPTION
**Pull request by qtran-n4**
Task: https://github.com/eclipse/n4js/issues/460

### CODE REVIEW

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java [hunk 1]
```diff
@@ -627,11 +627,40 @@ private boolean reduceParameterizedTypeRef(ParameterizedTypeRef left, Parameteri
...
-					wasAdded |= reduce(leftArg, leftParamSubst, variance.mult(INV));
+					final TypeVariable leftTypeVar = (left.getDeclaredType() != null
+							&& left.getDeclaredType().getTypeVars().size() > 0)
+									? left.getDeclaredType().getTypeVars().get(0) : null;
+					final TypeVariable rightTypeVar = (right.getDeclaredType() != null
+							&& right.getDeclaredType().getTypeVars().size() >= idx)
+									? right.getDeclaredType().getTypeVars().get(idx) : null;
+
+					boolean caseConvered = false;
+					if (variance == CO) {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java]: can this ever happen? (see normalization in lines 582-588, above)

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java [hunk 2]
```diff
@@ -627,11 +627,40 @@ private boolean reduceParameterizedTypeRef(ParameterizedTypeRef left, Parameteri
...
+							caseConvered = true;
+						} else if (rightTypeVar != null && rightTypeVar.isDeclaredContravariant()) {
+							wasAdded |= reduce(leftArg, leftParamSubst, CO);
+							caseConvered = true;
+						}
+					}
+					if (!caseConvered) {
+						wasAdded |= reduce(leftArg, leftParamSubst, variance.mult(INV));
+					}
+					// wasAdded |= reduce(leftArg, leftParamSubst, variance.mult(INV));
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java]: comment can be deleted

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java (hunk 1 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java]: Done. Remove this case.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java [hunk 3]
```diff
@@ -627,11 +627,39 @@ private boolean reduceParameterizedTypeRef(ParameterizedTypeRef left, Parameteri
...
 					}
-					wasAdded |= reduce(leftArg, leftParamSubst, variance.mult(INV));
+					final TypeVariable leftTypeVar = (left.getDeclaredType() != null
+							&& left.getDeclaredType().getTypeVars().size() > 0)
+									? left.getDeclaredType().getTypeVars().get(0) : null;
+					final TypeVariable rightTypeVar = (right.getDeclaredType() != null
+							&& right.getDeclaredType().getTypeVars().size() >= idx)
+									? right.getDeclaredType().getTypeVars().get(idx) : null;
+
+					boolean caseConvered = false;
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java]: type? do you mean `caseCovered`?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java (hunk 2 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java (hunk 3 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java]: Rewrote this and removed the flag.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java [hunk 4]
```diff
@@ -627,11 +627,39 @@ private boolean reduceParameterizedTypeRef(ParameterizedTypeRef left, Parameteri
 						throw new UnsupportedOperationException("unsupported subtype of TypeArgument: "
 								+ leftArg.getClass().getName());
 					}
-					wasAdded |= reduce(leftArg, leftParamSubst, variance.mult(INV));
+					final TypeVariable leftTypeVar = (left.getDeclaredType() != null
+							&& left.getDeclaredType().getTypeVars().size() > 0)
+									? left.getDeclaredType().getTypeVars().get(0) : null;
+					final TypeVariable rightTypeVar = (right.getDeclaredType() != null
+							&& right.getDeclaredType().getTypeVars().size() >= idx)
+									? right.getDeclaredType().getTypeVars().get(idx) : null;
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/constraints/Reducer.java]: there is a convenience method `Type#getVarianceOfTypeVar()` that might help simplify this a bit

**Comment by mor-n4** [comment]
please re-run builds.

If green, then :+1: to merge!

**Comment by qtran-n4** [comment]
@mor-n4 There was another problem with our changes and I fixed it. I will merge this PR for now but we should discuss that fix.

**Pull request merged**